test/lhash_test.c

```diff
@@ -0,0 +1,248 @@
+    }
+
+    /* replace */
+    j = 13;
+    if ((p = lh_int_insert(h, &j)) == NULL) {
+        fprintf(stderr, "test lhash int replacement insert\n");
+        goto end;
+    }
+    if (p != int_tests + 1) {
+        fprintf(stderr, "test lhash int replacement pointer\n");
+        goto end;
+    }
+    if (lh_int_retrieve(h, int_tests + 1) != &j) {
+        fprintf(stderr, "test lhash int replacement variable\n");
+        goto end;
+    }
+
+    /* do_all */
+    memset(int_found, 0, sizeof(int_found));
+    lh_int_doall(h, &int_doall);
+    for (i = 0; i < n_int_tests; i++)
+        if (int_found[i] != 1) {
+            fprintf(stderr, "test lhash int doall %d\n", i);
+            goto end;
+        }
+    
+    /* do_all_arg */
+    memset(int_found, 0, sizeof(int_found));
+    lh_int_doall_short(h, int_doall_arg, int_found);
+    for (i = 0; i < n_int_tests; i++)
+        if (int_found[i] != 1) {
+            fprintf(stderr, "test lhash int doall arg %d\n", i);
+            goto end;
+        }
+    
+    /* delete */
+    for (i = 0; i < n_dels; i++) {
+        const int b = lh_int_delete(h, &dels[i].data) == NULL;
+        if ((b ^ dels[i].null) != 0) {
+            fprintf(stderr, "test lhash int delete %d\n", i);
+            goto end;
+        }
+    }
+
+    /* error */
+    if (lh_int_error(h) != 0) {
+        fprintf(stderr, "test lhash int error\n");
+        goto end;
+    }
+
+    testresult = 1;
+end:
+    lh_int_free(h);
+    return testresult;
+}
+
+static unsigned long int stress_hash(const int *p)
+{
+    return *p;
+}
+
+static int test_stress(void)
+{
+    LHASH_OF(int) *h = lh_int_new(&stress_hash, &int_cmp);
+    const unsigned int n = 2500000;
+    unsigned int i;
+    int testresult = 0, *p;
+
+    if (h == NULL) {
+        fprintf(stderr, "test lhash stress allocation\n");
+        goto end;
+    }
+
+    /* insert */
+    for (i = 0; i < n; i++) {
+        p = malloc(sizeof(i));
+        if (p == NULL) {
+            fprintf(stderr, "test lhash stress out of memory %d\n", i);
+            goto end;
+        }
+        *p = 3 * i + 1;
+        lh_int_insert(h, p);
+    }
+
+    /* num_items */
+    if (lh_int_num_items(h) != n) {
+            fprintf(stderr, "test lhash stress num items\n");
+            goto end;
+    }
+
+    fprintf(stderr, "hash full statistics:\n");
+    OPENSSL_LH_stats((OPENSSL_LHASH *)h, stderr);
+    fprintf(stderr, "\nhash full node usage:\n");
+    OPENSSL_LH_node_usage_stats((OPENSSL_LHASH *)h, stderr);
+
+    /* delete in a different order */
+    for (i = 0; i < n; i++) {
+        const int j = (7 * i + 4) % n * 3 + 1;
+
+        if ((p = lh_int_delete(h, &j)) == NULL) {
+            fprintf(stderr, "test lhash stress delete %d\n", i);
+            goto end;
+        }
+        if (*p != j) {
+            fprintf(stderr, "test lhash stress bad value %d\n", i);
+            goto end;
+        }
+        free(p);
+    }
+
+    fprintf(stderr, "\nhash empty statistics:\n");
```

Comment:
There is a `TEST_check()`...  It's an assert that makes sure to print the OpenSSL error stack as well.

Generally, asserts are frowned upon unless *absolutely* necessary.  Even if one test fails, we do want to run the rest of them.